node/service/src/lib.rs

```diff
@@ -281,18 +298,40 @@ fn real_overseer<S: SpawnNamed>(
 }
 
 #[cfg(feature = "full-node")]
-fn new_full<RuntimeApi, Executor>(
+pub struct NewFull<C> {
+	pub task_manager: TaskManager,
+	pub client: C,
+	pub node_handles: OverseerHandler,
+	pub network: Arc<sc_network::NetworkService<Block, <Block as BlockT>::Hash>>,
+	pub network_status_sinks: service::NetworkStatusSinks<Block>,
+	pub rpc_handlers: RpcHandlers,
+}
+
+#[cfg(feature = "full-node")]
+impl<C> NewFull<C> {
+	fn with_client(self, func: impl FnOnce(C) -> Client) -> NewFull<Client> {
+		NewFull {
+			client: func(self.client),
+			task_manager: self.task_manager,
+			node_handles: self.node_handles,
+			network: self.network,
+			network_status_sinks: self.network_status_sinks,
+			rpc_handlers: self.rpc_handlers,
+		}
+	}
+}
+
+/// Create a new full node of arbitrary runtime and executor.
+///
+/// This is an advanced feature and not recommended for general use. Generally, `build_full` is
+/// a better choice.
+#[cfg(feature = "full-node")]
+pub fn new_full<RuntimeApi, Executor>(
```

Comment:
In the old service this also got a parameter "test" which was forgotten here.

Reply:
This should be brought back.

Reply:
Could you elaborate why is it needed. I see it used in two places:
https://github.com/paritytech/polkadot/blob/3b37ea68cd493d178e02161b8e2f907015ccad8c/service/src/lib.rs#L149-L154
https://github.com/paritytech/polkadot/blob/3b37ea68cd493d178e02161b8e2f907015ccad8c/service/src/lib.rs#L171-L175

Is it only to set the prometheus prefix or am I missing something?

Reply:
Maybe we can ignore it, I'm not sure. I don't have implemented this.

@cecton was this AFAIK.

Reply:
It's because the test service uses the test runtime and don't need prometheus.

But for the second place with kusama hard forks I think the condition `!test` can safely be removed because the test `config.chain_spec.is_kusama()` will be false for the test runtime anyway.

Originally there was one important other thing: the ValidationPool is built differently for the test node: https://github.com/paritytech/polkadot/blob/rococo-branch/service/src/lib.rs#L414

Reply:
> It's because the test service uses the test runtime and don't need prometheus.

but `if prometheus_config.is_none()`, this line will not be executed anyway, so it seems redundant

> Originally there was one important other thing: the ValidationPool is built differently for the test node: https://github.com/paritytech/polkadot/blob/rococo-branch/service/src/lib.rs#L414

do you think it's worth adding back to master or it can be taken care of in the rococo-branch?

Reply:
> do you think it's worth adding back to master or it can be taken care of in the rococo-branch?

I'm pretty sure rococo-branch is temporary (@bkchr correct me if I'm wrong). We will need to get it back or add a feature of some sort.

> but if prometheus_config.is_none(), this line will not be executed anyway, so it seems redundant

You're right!! I think it's because `new_full` used to be a macro so the code was compiled on cumulus where we don't have the dependency for that. Now it's not an issue anymore. :tada: 

:thinking: comes to think of it... it was the same thing for the kusama forks thingy.

Cool!! We can drop this safely